### PR TITLE
sssd: init at 1.14.2

### LINKED ISSUE #11407: Support SSSD
Add a package for the [System Security Services Daemon (SSSD)](https://fedorahosted.org/sssd/). Replace `nslcd` usage with this for the `users.ldap.daemon` NixOS options.

SSSD seems to be the more widely used and accepted option, now, with RHEL recommending it.

### PR DESCRIPTION
**Pull request by outergod**
###### Motivation for this change

This pull request is meant to supersede #14697 and fix #11407. I need sssd for work myself.

My additional changes include:
* patching glibc to look for NSS modules in `/run/nss-modules/$UNIQUE_ID_OF_GLIBC_`, first
* an `nss-modules` buildEnv package that includes all existing packages that used to refer to `system.nssModules`, plus sssd

I've ditched one of the original ideas to also support per-user NSS modules because I'm not sure how to meaningfully support them with the activation script for nss-modules. Additionally, it would require that numeric IDs don't change. Using the username would cause an infinite recursive loop when attempting to resolve the name, instead.

My system is still `nixos-rebuild --switch`ing for a full-fledged test, which will probably take really long. 

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by abbradar** [comment]
Is there any guarantees which glibc gives on ABI stability of NSS modules? If yes, we may want weaken path requirements (ideally it would be just `/run/nss-modules/` if ABI hasn't changed and is guaranteed to be forward-compatible, but we can do e.g. `/run/nss-modules/${glibc.version}/`). This is to ease pain when having packages compiled against different glibc versions in the system (e.g. right after the switch, or from `nix-env`).

File: pkgs/os-specific/linux/sssd/default.nix
Comment (abbradar) [inline pkgs/os-specific/linux/sssd/default.nix]: This and further those should be unnecesarry -- you don't change directory anywhere else.

File: pkgs/tools/misc/ding-libs/default.nix
Comment (abbradar) [inline pkgs/tools/misc/ding-libs/default.nix]: Same thing about `cd`.

File: pkgs/tools/text/sgml/jade/default.nix
Comment (abbradar) [inline pkgs/tools/text/sgml/jade/default.nix]: You can rather do:
```
  makeFlags = [ "DESTDIR=$(out)" ];
  preInstall = ''install -d -m755 "$out"/lib'';
  postInstall = ''mv "$out/bin/sx" "$out/bin/sgml2xml"'';
```
This way you'd reuse default `installPhase` which runs Make.

**Review by abbradar — CHANGES_REQUESTED** [review]
Generic packaging nitpicks. I'm not knowledgeable much about NSS in general and SSSD specifically, but overall I seem to understand what's going on and it seems good.

**Comment by abbradar** [comment]
It seems to me that ABI is stable: http://www.gnu.org/software/libc/manual/html_node/NSS-Basics.html . What do you think?

**Comment by e-user** [comment]
Do you suggest dumping X in `/run/nss-modules/$X` altogether or make it `2` for glibc's major version?

**Comment by abbradar** [comment]
Ideally the first, but I'm not sure at all about compatibility. I can't found any claims or guarantees. On the contrary, I found that nss-hostname [has an update for new glibc NSS interfaces](http://0pointer.de/lennart/projects/nss-myhostname/). I think we may want to ask around in [libc-help](https://www.gnu.org/software/libc/involved.html) about whether there are any de jure/de facto (historical) guarantees. Can you ask them and cc me (e-mail is in my profile)?

**Comment by Mic92** [comment]
This is useful for other nss modules. Currently nss-myhostname is not working

File: pkgs/os-specific/linux/sssd/default.nix
Comment (e-user) [inline pkgs/os-specific/linux/sssd/default.nix]: To be honest, I didn't even look at @benwbooth's original submission properly. I'm using this opportunity to clean up all of it. Do you think it's a good idea to update to the latest SSSD before finalizing the submission? There seem to have been a couple of releases ever since.

**Comment by e-user** [comment]
Done, email sent.

**Comment by edolstra** [comment]
I'd prefer not to have `/run/nss-modules`. The way that NixOS supports NSS modules is via nscd. That way, NSS modules only need to be in the `LD_LIBRARY_PATH` of nscd. This also ensures that they work for 32-bit applications on 64-bit systems.

**Comment by e-user** [comment]
Hi @edolstra, may I point your attention to this bit from the [RHEL documentation](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/System-Level_Authentication_Guide/usingnscd-sssd.html):
```
SSSD is not designed to be used with the NSCD daemon. Even though SSSD does not directly conflict 
with NSCD, using both services can result in unexpected behavior, especially with how long entries are 
cached.
```

nscd'd caching actually breaks the way SSSD works, which should always have precedence and does its own caching, AFAIK. From what I can collect on the net, SSSD is also meant to replace nscd, eventually.

**Comment by Mic92** [comment]
@e-user could SSSD make use of LD_LIBRARY_PATH?

**Comment by e-user** [comment]
I can try to build SSSD so that it is only used through nscd and LD_LIBRARY_PATH. Maybe switching off nscd's caching for everything but host name resolution does the trick.

**Comment by Mic92** [comment]
@edolstra btw, the libnss modules of systemd are not used by nscd, because LD_LIBRARY_PATH points to the lib output, while the nss modules live in the systemd default out package.

**Comment by e-user** [comment]
This happens when attempting to start SSSD with nscd present:
```
sssd[26627]: NSCD socket was detected. NSCD caching capabilities may conflict with SSSD for users and groups. It is recommended not to run NSCD in parallel with SSSD, unless NSCD is configured not to cache the passwd, group, netgroup and services nsswitch maps.
```

Investigating.

**Comment by abbradar** [comment]
If `nscd` abstracts modules over a socket, that indeed seems to be a superior solution. If it can be configured to not conflict, avoiding all the hassle of ABI compatibility is awesome. Let's hope it works!

**Comment by e-user** [comment]
I agree. It seems like SSSD will stop complaining if nscd is configured not to cache the services it provides.  That means nscd's config needs to change when SSSD is enabled. I'm working on it.

**Comment by e-user** [comment]
Great news: I set SSSD's `debug_level` to 9 and ran a `getent passwd` query with nscd up - and could see it was attempting to resolve the request.

**Comment by e-user** [comment]
I force-pushed a new branch that is cleansed from the previous mess and updates ding-libs to 0.6.0 and SSSD to the current stable release, 1.14.2. The most severe change is to how the NixOS nscd and SSSD modules are configured and interact, which will probably require a careful review.  
What's still left to do is adding an option to supply SSSD with configuration, insert the default configuration without which SSSD will refuse to run and figure out how to enable its [DBUS-specific bits](https://fedorahosted.org/sssd/wiki/DesignDocs/DBusResponder). Also, there seems to be a PAM module whose purpose I'm not sure of, yet - but it seems to be required by the default configuration. I'm also not sure how and whether to tackle the Kerberos plugins SSSD provides.

File: nixos/modules/services/misc/sssd.nix
Comment (abbradar) [inline nixos/modules/services/misc/sssd.nix]: I think we shouldn't use `mkForce`; rather, add an `assertion` that it's enabled. It is anyway by default but this way if user disables it for some reason they'd know of problems explicitly and can choose what to do.

File: pkgs/os-specific/linux/sssd/default.nix
Comment (abbradar) [inline pkgs/os-specific/linux/sssd/default.nix]: Ugh, why does it need nscd config? We can symlink it into `/etc` to make it accessible. Otherwise people need to rebuild SSSD each time they change config, which is not nice.

**Review by abbradar — CHANGES_REQUESTED** [review]
Commented on several things.

If the package has dbus .service files you can enable them by adding `sssd` to `services.dbus.packages`. But I don't think it needs to have them, daemons can just register themselves in D-Bus when they start (should work out of the box).

Also let's just add PAM modules when SSSD is enabled. Not sure what to do about Kerberos... I don't think we have any module system for it right now (disclaimer: I know barely a thing about Kerberos).

File: nixos/modules/services/misc/sssd.nix
Comment (e-user) [inline nixos/modules/services/misc/sssd.nix]: Good point, I didn't know that.

File: pkgs/os-specific/linux/sssd/default.nix
Comment (e-user) [inline pkgs/os-specific/linux/sssd/default.nix]: Ha, I was exactly asking myself what the _proper_ way to do this would be - so I let nscd create a symlink for its current configuration file in `/etc` and just point SSSD at its absolute path..?

I can elaborate on the SSSD-nscd specific bit. SSSD queries whether nscd is running (`--with-nscd`) and will additionally inspect its active configuration file (`--with-nscd-conf`). If it's not running, it won't do anything. If it is running and the configuration file doesn't disable passwd,  group, netgroup and services caching, it will print a warning to syslog because it essentially breaks SSSD. 

In case you're interested about the details, it's defined in `src/monitor.c`:
```c
    /* Warn if nscd seems to be running */
    ret = check_file(NSCD_SOCKET_PATH,
                     -1, -1, S_IFSOCK, S_IFMT, NULL, false);
    if (ret == EOK) {
        ret = sss_nscd_parse_conf(NSCD_CONF_PATH);

        switch (ret) {
            case ENOENT:
                sss_log(SSS_LOG_NOTICE,
                        "NSCD socket was detected. NSCD caching capabilities "
                        "may conflict with SSSD for users and groups. It is "
                        "recommended not to run NSCD in parallel with SSSD, "
                        "unless NSCD is configured not to cache the passwd, "
                        "group, netgroup and services nsswitch maps.");
                break;

            case EEXIST:
                sss_log(SSS_LOG_NOTICE,
                        "NSCD socket was detected and seems to be configured "
                        "to cache some of the databases controlled by "
                        "SSSD [fpasswd,group,netgroup,services]. It is "
                        "recommended not to run NSCD in parallel with SSSD, "
                        "unless NSCD is configured not to cache these.");
                break;

            case EOK:
                DEBUG(SSSDBG_TRACE_FUNC, "NSCD socket was detected and it "
                            "seems to be configured not to interfere with "
                            "SSSD's caching capabilities\n");
```

**Comment by abbradar** [comment]
Thank you for elaborated response! I see now; well, let's then indeed link it to `/etc`.  You can do it with `environment.etc`; let's then point nscd there too, just for consistency. Don't forget `restartTriggers` so that systemd services will be restarted when configuration file changes. 

On December 16, 2016 1:25:47 PM GMT+03:00, Alexander Kahl <notifications@github.com> wrote:
>e-user commented on this pull request.
>
>
>
>> @@ -39,7 +41,11 @@ stdenv.mkDerivation {
>       --without-semanage
>       --with-xml-catalog-path=''${SGML_CATALOG_FILES%%:*}
>       --with-ldb-lib-dir=$out/modules/ldb
>-      --with-nscd=${glibc.bin}/sbin/nscd
>+      ${lib.optionalString (nscdConfig != null) ''
>+        --with-nscd=${glibc.bin}/sbin/nscd
>+        --with-nscd-conf=${nscdConfig}
>
>Ha, I was exactly asking myself what the _proper_ way to do this would
>be - so I let nscd create a symlink for its current configuration file
>in `/etc` and just point SSSD at its absolute path..?
>
>I can elaborate on the SSSD-nscd specific bit. SSSD queries whether
>nscd is running (`--with-nscd`) and will additionally inspect its
>active configuration file (`--with-nscd-conf`). If it's not running, it
>won't do anything. If it is running and the configuration file doesn't
>disable passwd,  group, netgroup and services caching, it will print a
>warning to syslog because it essentially breaks SSSD. 
>
>In case you're interested about the details, it's defined in
>`src/monitor.c`:
>```c
>    /* Warn if nscd seems to be running */
>    ret = check_file(NSCD_SOCKET_PATH,
>                     -1, -1, S_IFSOCK, S_IFMT, NULL, false);
>    if (ret == EOK) {
>        ret = sss_nscd_parse_conf(NSCD_CONF_PATH);
>
>        switch (ret) {
>            case ENOENT:
>                sss_log(SSS_LOG_NOTICE,
>                 "NSCD socket was detected. NSCD caching capabilities "
>                  "may conflict with SSSD for users and groups. It is "
>                  "recommended not to run NSCD in parallel with SSSD, "
>                  "unless NSCD is configured not to cache the passwd, "
>                        "group, netgroup and services nsswitch maps.");
>                break;
>
>            case EEXIST:
>                sss_log(SSS_LOG_NOTICE,
>                 "NSCD socket was detected and seems to be configured "
>                        "to cache some of the databases controlled by "
>                       "SSSD [fpasswd,group,netgroup,services]. It is "
>                  "recommended not to run NSCD in parallel with SSSD, "
>                      "unless NSCD is configured not to cache these.");
>                break;
>
>            case EOK:
>            DEBUG(SSSDBG_TRACE_FUNC, "NSCD socket was detected and it "
>                        "seems to be configured not to interfere with "
>                            "SSSD's caching capabilities\n");
>```
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/21150

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cruegge** [comment]
Concerning PAM: I think the `pam_sss` module replaces `pam_krb5` and `pam_ccred`, so all `krb5`-stuff in `pam.nix` should probably be conditional on `config.krb5.enable && !config.sssd.enable`. I'm not sure whether enabling `sssd` should automatically enable `krb5`, though, i.e. is `krb5Full` required for `sssd`?

File: nixos/modules/services/misc/sssd.nix
Comment (abbradar) [inline nixos/modules/services/misc/sssd.nix]: This will keep the problem because `source` points to the changing path. Instead I'd drop `nscdConfig` argument and just hardcode it to `/etc/nscd.conf` in the package.. This way you'd also get rid of an override.

File: nixos/modules/security/pam.nix
Comment (abbradar) [inline nixos/modules/security/pam.nix]: Nitpick: use `optional`.

File: nixos/modules/services/system/nscd.nix
Comment (abbradar) [inline nixos/modules/services/system/nscd.nix]: We can now just omit `-f` because `/etc/nscd.conf` is the default path.

File: nixos/modules/services/system/nscd.nix
Comment (abbradar) [inline nixos/modules/services/system/nscd.nix]: Same thing about `-f`.

**Review by abbradar — CHANGES_REQUESTED** [review]
New wave of comments (sorry for being persistent, I think we are getting close!).

Questions about PAM and KRB are out of my league, but I think we can rely on you there given that you actually use it ~_^.

File: pkgs/os-specific/linux/sssd/default.nix
Comment (abbradar) [inline pkgs/os-specific/linux/sssd/default.nix]: Why call `make` by hand? Instead I'd move all those flags to `makeFlags` and use `postInstall` -- you can use `$out` there as `$(out)`.

File: pkgs/tools/text/sgml/jade/default.nix
Comment (abbradar) [inline pkgs/tools/text/sgml/jade/default.nix]: Instead: `NIX_CFLAGS_COMPILE = [ "-fpermissive" "-Wno-deprecated" ];` -- let's not use Bash when we can rely on Nix. BTW `-fpermissive` allows pretty scary things like calling undefined functions IIRC -- why is it needed?

**Review by abbradar — CHANGES_REQUESTED** [review]
More comments for the whole patch -- sorry, I should have done it in one go.

File: pkgs/tools/text/sgml/jade/default.nix
Comment (abbradar) [inline pkgs/tools/text/sgml/jade/default.nix]: Oh, only now I have noticed that not all changes are yours! Still, are you okay with fixing those?

**Review by abbradar — APPROVED** [review]

**Comment by abbradar** [comment]
This seems good to me now. It doesn't change anything seriously if you don't have SSSD enabled so I'll wait for a few days and merge this -- if noone else has complaints. Thank you for your work!

**Comment by abbradar** [comment]
Oh, one more thing -- can you please squash commits so that review changes would not be separate ones? Generally we have one commit per either new package/module/meaningful change, but you also have another author's commits there so let's leave them in pristine state and add patches on top.

**Comment by dezgeg** [comment]
Ah, darn, I was just working on the same stuff last week. I will take a look shortly. FWIW here's what I ended up with: https://github.com/dezgeg/nixpkgs/tree/sssd

File: pkgs/os-specific/linux/sssd/default.nix
Comment (dezgeg) [inline pkgs/os-specific/linux/sssd/default.nix]: This version here won't be visible to `nix-env -q`, it needs to be e.g. `name = "sssd-${version}";`.

File: pkgs/os-specific/linux/sssd/default.nix
Comment (dezgeg) [inline pkgs/os-specific/linux/sssd/default.nix]: `meta.version` is not needed.

**Comment by e-user** [comment]
Unfortunately, we're not done yet. E.g. there is no mechanism to provide SSSD's own configuration, yet. Crucial PAM bits are missing. @dezgeg provided some bits that are missing in my implementation as well, like `authorized_keys_command`.   
Since I don't have an IPA instance to test all of SSSD with at home, I'll create one in AWS.  
Please bear with me, @abbradar.

@dezgeg, do you also need SSSD for IPA, or just as-is?

File: nixos/modules/services/misc/nscd-sssd.conf
Comment (e-user) [inline nixos/modules/services/misc/nscd-sssd.conf]: I don't get any errors and I can see SSSD processing the request when tracing is switched on - please see above.

**Comment by abbradar** [comment]
No problem, I have misunderstood that it was done already. Unfortunately I don't have any good pieces of advice to give on SSSD itself.

**Comment by dezgeg** [comment]
I used the LDAP backend. PAM has worked fine for me so far, I don't think I changed anything there from the original commits.

File: nixos/modules/services/misc/nscd-sssd.conf
Comment (dezgeg) [inline nixos/modules/services/misc/nscd-sssd.conf]: What happened to me was after PAM login e.g. bash would set my prompt to `I have no name!@nix-vm $` and `whoami` wouldn't work. Also according to glibc sources that seems the intended behaviour:

````
1155       /* Is this service enabled?  */
1156       if (__glibc_unlikely (!db->enabled))
1157         {
1158           /* No, sent the prepared record.  */
1159           if (TEMP_FAILURE_RETRY (send (fd, db->disabled_iov->iov_base,
1160                                         db->disabled_iov->iov_len,
1161                                         MSG_NOSIGNAL))
1162               != (ssize_t) db->disabled_iov->iov_len
1163               && __builtin_expect (debug_level, 0) > 0)
1164             {
1165               /* We have problems sending the result.  */
1166               char buf[256];
1167               dbg_log (_("cannot write result: %s"),
1168                        strerror_r (errno, buf, sizeof (buf)));
1169             }
1170 
1171           return;
1172         }
````
... with `db->enabled` set as:
````
143       else if (strcmp (entry, "enable-cache") == 0)
144         {
145           int idx = find_db (arg1);
146           if (idx >= 0)
147             {
148               if (strcmp (arg2, "no") == 0)
149                 dbs[idx].enabled = 0;
150               else if (strcmp (arg2, "yes") == 0)
151                 dbs[idx].enabled = 1;
152             }
153         }
````

**Comment by e-user** [comment]
After messing around with sssd.conf and krb5.conf for a bit, I was finally able to log into NixOS. However, it seems like SSH is the only thing using SSSD right now, resolving users doesn't seem to work. Investigating.

**Comment by dezgeg** [comment]
BTW, another thing that I had to debug for quite a while was that without setting `services.openssh.challengeResponseAuthentication = false;` SSH wouldn't work - it was giving a non-plaintext password to the PAM layer.

**Comment by e-user** [comment]
Thanks @dezgeg, for some reason, it works without setting that, for me.  
But you were right about another thing: If nscd is configured to not cache users and groups, it actually won't even try to query SSSD for them. This is a little bit annoying as it means we have to create bogus caching on and eat SSSD's bootup error message, unless we patch it. *sigh*

**Comment by e-user** [comment]
@edolstra as it seems, nscd only forwards requests for services it supports and caching is switched on, for.  We have to be aware that using nscd for all NSS modules is essentially a hack. Since nscd only support `passwd, group, hosts, services, or netgroup` according to `man 5 nscd.conf`, it means NSS modules for networks, ethers, protocols and automount (which affects SSSD) still won't be supported this way.

**Comment by e-user** [comment]
I took over some of your additions, @dezgeg. However, for some weird reason, `security.pam.makeHomeDir = true;` doesn't actually add the string to the PAM configuration files for me, on two different NixOS installations. Could you please help me find out what's wrong?

**Comment by dezgeg** [comment]
I added this:
````
security.pam.services.su.makeHomeDir = true;
security.pam.services.sshd.makeHomeDir = true;
````

I don't know what's the way to change it for all modules, though it would be better obviously.

**Comment by Mic92** [comment]
@e-user one option I see is to disable nscd, when sssd is enabled. Would that work?

**Comment by e-user** [comment]
@Mic92 unfortunately and quite paradoxically that will break SSSD. @edolstra doesn't like the original approach to load NSS modules from a fixed location to circumvent the circular sssd->glibc->sssd dependency problem, so we're essentially abusing nscd as a no-cache module loader. I'm not 100% happy with it, but it does actually work.

**Comment by e-user** [comment]
I'll rebase once more and squash the commits as promised. @dezgeg, would you like to become co-maintainer of SSSD and its new dependencies?

**Comment by abbradar** [comment]
Given that our current way of supporting NSS modules can't get full SSSD support (nscd only caches particular types of databases, so you don't get for example automount of shared corporate storages if I understand correctly) -- @edolstra, what do you think of moving to `/run/nss-modules{,-32}`? On one hand, this gives us full support and ABI risks are minimal according to [reply in libc-help](http://sourceware.org/ml/libc-help/2016-12/msg00006.html). On the other hand, the libraries can fail to load because of address space conflicts as evidenced [in another reply](http://sourceware.org/ml/libc-help/2016-12/msg00008.html) and crash the binary they are loaded into (but crashing the daemon is arguably not much better) if written badly.

I personally don't have a strong opinion (arguments on both sides are strong and understandable) but we may want to review that once more in light of new problems.

**Comment by e-user** [comment]
I'm not sure whether waiting for @edolstra is worth it. I have to keep rebasing my changes and fixing conflicts while others and myself are waiting for this feature, so it might be easier to merge the changes and let me file a bug right afterwards that explains how supporting external NSS modules through nscd is a hack and lists things that don't work correctly.

**Comment by abbradar** [comment]
Indeed, moving to the new way of handling NSS is of a bigger scope than this PR. I don't have time to run tests for this (even with SSSD disabled, just to make sure nothing is broken) but may have more time at the weekend. If someone beats me to it I don't mind ~_^.

**Comment by e-user** [comment]
Hey @abbradar, any news?

**Comment by abbradar** [comment]
Thanks for the reminder, life caught on. I'll look at this today. 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by abbradar** [comment]
Looks like `jade` fails to build now:
```
/nix/store/p9g0vm6bj88sdknpjw9gwziyh8hjw71g-binutils-2.27/bin/ld: cannot find -lsp
/nix/store/p9g0vm6bj88sdknpjw9gwziyh8hjw71g-binutils-2.27/bin/ld: cannot find -lspgrove
```
This is atop of current unstable channel. Any ideas?

**Comment by dezgeg** [comment]
Still on the `/run/nss-modules` vs nscd discussion: It's not just the `libnss` ABI stability that needs consideration but also any dependent shared libraries of the NSS module itself. E.g. `libnss_hostname.so` has this:

````
	libcap.so.2 => /nix/store/73zh7aw558bdahk27j4l825wiz46p1g8-libcap-2.25-lib/lib/libcap.so.2 (0x00007f6fe145d000)
````

I don't know if it applies to `__libc_dlopen` used by glibc when loading the NSS module, but at least in some other cases (https://github.com/NixOS/nixpkgs/issues/16779) impurely loaded shared library dependencies have been incompatible with a different version of the corresponding library causing really weird failure modes.

Maybe it's worth asking the glibc folks for a better way of doing this non-caching passthru with nscd (and support for automount maps). I'd imagine other distros would find this use case useful as well.

**Comment by e-user** [comment]
Using my own machine with its extensive system package collection is too time-consuming for master build cycles over and over again, so I'll start using some AWS EC2 power on Monday.

@dezgeg @abbradar how are you guys' C programming skills? It might be worth looking into extending nscd. The objective would be to a) add the missing services and b) an additional passthru option for each of the services.

**Comment by abbradar** [comment]
@e-user You can build on the top of [latest unstable channel](https://github.com/NixOS/nixpkgs-channels), this should decrease the build times drastically. Testing it on unstable channel is good enough, I believe.

I feel confident to write C but not confident enough to write complex and secure C (given that `nscd` runs as root and is used to authenticate users!). However it depends on how complex the code to add this to nscd would be. I imagine that if it's written nicely a simple addition of `case`s/`if`s with duplicated code from other tested paths would suffice, but my previous encounter with code in glibc (namely ld.so) wasn't so nice D:

**Comment by e-user** [comment]
@abbradar I just built jade successfully atop `9d8a179`, which is latest in `channels/nixos-unstable` at the time of writing. No changes made. I just pushed the rebased version.

I'll take a look at nscd. Let's see whether its maintainers would want such a change in to begin with.

**Comment by e-user** [comment]
Good news regarding sudoers and automount in nsswitch.conf: AFAIK from their respective documentation and glibc's source code, glibc doesn't implement queries for those itself but the respective services read nsswitch.conf autonomously and query NSS modules themselves. This also means that LD_LIBRARY_PATH have to be extended for them, correspondingly. 

@abbradar, would you like me to also add this for this pull request or create a subsequent ticket that reports the missing behavior as a bug? Also, is there anything else preventing a merge right now? Would you like me to provide documentation? I've never provided NixOS documentation before, are there any pointers how this works?

**Comment by abbradar** [comment]
It still fails for me -- do you have `nix.useSandbox = true;` in your NixOS configuration?

Apart from that I think nothing holds this from being merged -- you can leave improvements for subsequent PRs.

**Comment by e-user** [comment]
Nope, sorry, still works with sandboxing on. Could you please try in a nixos-container or VM?

**Comment by abbradar** [comment]
After testing on other machines I found the reason -- `jade` Makefiles are not parallel building safe. If it's disabled it works for me. For reference: I have 8 build cores.

I went ahead and merged it with this change. Thank you!

**Comment by e-user** [comment]
Thank you @abbradar for your patience and persistence!

**Pull request closed without merging**
